File: VB/CodeUtils/Evaluator.vb
Imports System
Imports System.CodeDom.Compiler
Imports System.Reflection

Namespace RichEditDocumentServerAPIExample.CodeUtils

    Public MustInherit Class ExampleCodeEvaluator

        Protected MustOverride ReadOnly Property CodeStart As String

        Protected MustOverride ReadOnly Property CodeBeforeClasses As String

        Protected MustOverride ReadOnly Property CodeEnd As String

        Protected MustOverride Function GetCodeDomProvider() As CodeDomProvider

        Protected MustOverride Function GetModuleAssembly() As String

        Protected MustOverride Function GetExampleClassName() As String

        Public Function ExecuteCodeAndGenerateDocument(ByVal args As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs) As Boolean
            Dim theCode As String = System.[String].Concat(Me.CodeStart, args.Code, Me.CodeBeforeClasses, args.CodeClasses, Me.CodeEnd)
            Dim linesOfCode As String() = New String() {theCode}
            Return Me.CompileAndRun(linesOfCode, args.EvaluationParameter)
        End Function

        Protected Friend Function CompileAndRun(ByVal linesOfCode As String(), ByVal evaluationParameter As Object) As Boolean
            Dim CompilerParams As System.CodeDom.Compiler.CompilerParameters = New System.CodeDom.Compiler.CompilerParameters()
            CompilerParams.GenerateInMemory = True
            CompilerParams.TreatWarningsAsErrors = False
            CompilerParams.GenerateExecutable = False
            Dim referencesSystem As String() = New String() {"System.dll", "System.Windows.Forms.dll", "System.Data.dll", "System.Xml.dll", "System.Core.dll", "System.Drawing.dll"}
            Dim referencesDX As String() = New String() {AssemblyInfo.SRAssemblyData, Me.GetModuleAssembly(), AssemblyInfo.SRAssemblyOfficeCore, AssemblyInfo.SRAssemblyPrintingCore, AssemblyInfo.SRAssemblyPrinting, AssemblyInfo.SRAssemblyDocs, AssemblyInfo.SRAssemblyUtils, AssemblyInfo.SRAssemblyRichEdit}
            Dim references As String() = New String(referencesSystem.Length + referencesDX.Length - 1) {}
            For referenceIndex As Integer = 0 To referencesSystem.Length - 1
                references(referenceIndex) = referencesSystem(referenceIndex)
            Next

            Dim i As Integer = 0, initial As Integer = referencesSystem.Length
            While i < referencesDX.Length
                Dim assembly As System.Reflection.Assembly = System.Reflection.Assembly.Load(referencesDX(i) & AssemblyInfo.FullAssemblyVersionExtension)
                If assembly IsNot Nothing Then references(i + initial) = assembly.Location
                i += 1
            End While

            CompilerParams.ReferencedAssemblies.AddRange(references)
            Dim provider As System.CodeDom.Compiler.CodeDomProvider = Me.GetCodeDomProvider()
            Dim compile As System.CodeDom.Compiler.CompilerResults = provider.CompileAssemblyFromSource(CompilerParams, linesOfCode)
            If compile.Errors.HasErrors Then
                Dim text As String = "Compile error: "
                For Each ce As System.CodeDom.Compiler.CompilerError In compile.Errors
                    text += "rn" & ce.ToString()
                Next

                System.Windows.Forms.MessageBox.Show(text)
                Return False
            End If

            Dim [module] As System.Reflection.[Module] = Nothing
            Try
                [module] = compile.CompiledAssembly.GetModules()(0)
            Catch
            End Try

            Dim moduleType As System.Type = Nothing
            If [module] Is Nothing Then
                Return False
            End If

            moduleType = [module].[GetType](Me.GetExampleClassName())
            Dim methInfo As System.Reflection.MethodInfo = Nothing
            If moduleType Is Nothing Then
                Return False
            End If

            methInfo = moduleType.GetMethod("Process")
            If methInfo IsNot Nothing Then
                Try
                    methInfo.Invoke(Nothing, New Object() {evaluationParameter})
                Catch __unusedException1__ As System.Exception
                    Return False ' an error
                End Try

                Return True
            End If

            Return False
        End Function
    End Class

    Public MustInherit Class RichEditExampleCodeEvaluator
        Inherits RichEditDocumentServerAPIExample.CodeUtils.ExampleCodeEvaluator

        Protected Overrides Function GetModuleAssembly() As String
            Return AssemblyInfo.SRAssemblyRichEditCore
        End Function

        Protected Overrides Function GetExampleClassName() As String
            Return "RichEditCodeResultViewer.ExampleItem"
        End Function
    End Class

#Region "RichEditCSExampleCodeEvaluator"
    Public Class RichEditCSExampleCodeEvaluator
        Inherits RichEditDocumentServerAPIExample.CodeUtils.RichEditExampleCodeEvaluator

        Protected Overrides Function GetCodeDomProvider() As CodeDomProvider
            Return New Microsoft.CSharp.CSharpCodeProvider()
        End Function

        Const codeStartField As String = "using System;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Text;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using DevExpress.XtraPrinting;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using DevExpress.XtraRichEdit;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using DevExpress.XtraRichEdit.API.Native;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using DevExpress.XtraRichEdit.Export;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using DevExpress.XtraRichEdit.Import;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using DevExpress.Office.Utils;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Drawing;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Windows.Forms;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using DevExpress.Utils;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.IO;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Diagnostics;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Xml;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Data;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Collections.Generic;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Linq;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using System.Globalization;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using Document = DevExpress.XtraRichEdit.API.Native.Document;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "using TableRow = DevExpress.XtraRichEdit.API.Native.TableRow;" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "namespace RichEditCodeResultViewer { " & Global.Microsoft.VisualBasic.Constants.vbCrLf & "public class ExampleItem { " & Global.Microsoft.VisualBasic.Constants.vbCrLf & "        public static void Process(RichEditDocumentServer wordProcessor) { " & Global.Microsoft.VisualBasic.Constants.vbCrLf & Global.Microsoft.VisualBasic.Constants.vbCrLf

        Const codeBeforeClassesField As String = "       " & Global.Microsoft.VisualBasic.Constants.vbCrLf & " }" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "    }" & Global.Microsoft.VisualBasic.Constants.vbCrLf

        Const codeEndField As String = Global.Microsoft.VisualBasic.Constants.vbCrLf & "    }" & Global.Microsoft.VisualBasic.Constants.vbCrLf

        Protected Overrides ReadOnly Property CodeStart As String
            Get
                Return RichEditDocumentServerAPIExample.CodeUtils.RichEditCSExampleCodeEvaluator.codeStartField
            End Get
        End Property

        Protected Overrides ReadOnly Property CodeBeforeClasses As String
            Get
                Return RichEditDocumentServerAPIExample.CodeUtils.RichEditCSExampleCodeEvaluator.codeBeforeClassesField
            End Get
        End Property

        Protected Overrides ReadOnly Property CodeEnd As String
            Get
                Return RichEditDocumentServerAPIExample.CodeUtils.RichEditCSExampleCodeEvaluator.codeEndField
            End Get
        End Property
    End Class

#End Region
#Region "RichEditVbExampleCodeEvaluator"
    Public Class RichEditVbExampleCodeEvaluator
        Inherits RichEditDocumentServerAPIExample.CodeUtils.RichEditExampleCodeEvaluator

        Protected Overrides Function GetCodeDomProvider() As CodeDomProvider
            Return New Microsoft.VisualBasic.VBCodeProvider()
        End Function

        Const codeStartField As String = "Imports Microsoft.VisualBasic" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports DevExpress.XtraRichEdit" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports DevExpress.XtraRichEdit.API.Native" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.Drawing" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.Windows.Forms" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports DevExpress.Utils" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.IO" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.Diagnostics" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.Xml" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.Data" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.Collections.Generic" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports System.Globalization" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports Document = DevExpress.XtraRichEdit.API.Native.Document" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Imports TableRow = DevExpress.XtraRichEdit.API.Native.TableRow" & Global.Microsoft.VisualBasic.Constants.vbCrLf & "Namespace RichEditCodeResultViewer" & Global.Microsoft.VisualBasic.Constants.vbCrLf & Global.Microsoft.VisualBasic.Constants.vbTab & "Public Class ExampleItem" & Global.Microsoft.VisualBasic.Constants.vbCrLf & Global.Microsoft.VisualBasic.Constants.vbTab & Global.Microsoft.VisualBasic.Constants.vbTab & "Public Shared Sub Process(ByVal wordProcessor As RichEditDocumentServer)" & Global.Microsoft.VisualBasic.Constants.vbCrLf & Global.Microsoft.VisualBasic.Constants.vbCrLf

        Const codeBeforeClassesField As String = Global.Microsoft.VisualBasic.Constants.vbCrLf & Global.Microsoft.VisualBasic.Constants.vbTab & Global.Microsoft.VisualBasic.Constants.vbTab & "End Sub" & Global.Microsoft.VisualBasic.Constants.vbCrLf & Global.Microsoft.VisualBasic.Constants.vbTab & "End Class" & Global.Microsoft.VisualBasic.Constants.vbCrLf

        Const codeEndField As String = Global.Microsoft.VisualBasic.Constants.vbCrLf & "End Namespace" & Global.Microsoft.VisualBasic.Constants.vbCrLf

        Protected Overrides ReadOnly Property CodeStart As String
            Get
                Return RichEditDocumentServerAPIExample.CodeUtils.RichEditVbExampleCodeEvaluator.codeStartField
            End Get
        End Property

        Protected Overrides ReadOnly Property CodeBeforeClasses As String
            Get
                Return RichEditDocumentServerAPIExample.CodeUtils.RichEditVbExampleCodeEvaluator.codeBeforeClassesField
            End Get
        End Property

        Protected Overrides ReadOnly Property CodeEnd As String
            Get
                Return RichEditDocumentServerAPIExample.CodeUtils.RichEditVbExampleCodeEvaluator.codeEndField
            End Get
        End Property
    End Class

#End Region
    Public MustInherit Class ExampleEvaluatorByTimer
        Implements System.IDisposable

        Private leakSafeCompileEventRouter As RichEditDocumentServerAPIExample.CodeUtils.LeakSafeCompileEventRouter

        Private compileExampleTimer As System.Windows.Forms.Timer

        Private compileComplete As Boolean = True

        Const CompileTimeIntervalInMilliseconds As Integer = 2000

        Public Sub New(ByVal enableTimer As Boolean)
            Me.leakSafeCompileEventRouter = New RichEditDocumentServerAPIExample.CodeUtils.LeakSafeCompileEventRouter(Me)
            If enableTimer Then
                Me.compileExampleTimer = New System.Windows.Forms.Timer()
                Me.compileExampleTimer.Interval = RichEditDocumentServerAPIExample.CodeUtils.ExampleEvaluatorByTimer.CompileTimeIntervalInMilliseconds
                AddHandler Me.compileExampleTimer.Tick, New System.EventHandler(AddressOf Me.leakSafeCompileEventRouter.OnCompileExampleTimerTick) 'OnCompileTimerTick
                Me.compileExampleTimer.Enabled = True
            End If
        End Sub

        Public Sub New()
            Me.New(True)
        End Sub

#Region "Events"
        Public Event QueryEvaluate As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventHandler

        Protected Friend Overridable Function RaiseQueryEvaluate() As CodeEvaluationEventArgs
            If QueryEvaluateEvent IsNot Nothing Then
                Dim args As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs = New RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs()
                RaiseEvent QueryEvaluate(Me, args)
                Return args
            End If

            Return Nothing
        End Function

        Public Event OnBeforeCompile As System.EventHandler

        Private Sub RaiseOnBeforeCompile()
            RaiseEvent OnBeforeCompile(Me, New System.EventArgs())
        End Sub

        Public Event OnAfterCompile As RichEditDocumentServerAPIExample.CodeUtils.OnAfterCompileEventHandler

        Private Sub RaiseOnAfterCompile(ByVal result As Boolean)
            RaiseEvent OnAfterCompile(Me, New RichEditDocumentServerAPIExample.CodeUtils.OnAfterCompileEventArgs() With {.Result = result})
        End Sub

#End Region
        Public Sub CompileExample(ByVal sender As Object, ByVal e As System.EventArgs)
            If Not Me.compileComplete Then Return
            Dim args As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs = Me.RaiseQueryEvaluate()
            If Not args.Result Then Return
            Me.ForceCompile(args)
        End Sub

        Public Sub ForceCompile(ByVal args As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs)
            Me.compileComplete = False
            If Not System.[String].IsNullOrEmpty(args.Code) Then Me.CompileExampleAndShowPrintPreview(args)
            Me.compileComplete = True
        End Sub

        Private Sub CompileExampleAndShowPrintPreview(ByVal args As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs)
            Dim evaluationSucceed As Boolean = False
            Try
                Me.RaiseOnBeforeCompile()
                evaluationSucceed = Me.Evaluate(args)
            Finally
                Me.RaiseOnAfterCompile(evaluationSucceed)
            End Try
        End Sub

        Public Function Evaluate(ByVal args As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs) As Boolean
            Dim richeditExampleCodeEvaluator As RichEditDocumentServerAPIExample.CodeUtils.ExampleCodeEvaluator = Me.GetExampleCodeEvaluator(args.Language)
            Return richeditExampleCodeEvaluator.ExecuteCodeAndGenerateDocument(args)
        End Function

        Protected MustOverride Function GetExampleCodeEvaluator(ByVal language As RichEditDocumentServerAPIExample.CodeUtils.ExampleLanguage) As ExampleCodeEvaluator

        Public Sub Dispose() Implements Global.System.IDisposable.Dispose
            If Me.compileExampleTimer IsNot Nothing Then
                Me.compileExampleTimer.Enabled = False
                If Me.leakSafeCompileEventRouter IsNot Nothing Then RemoveHandler Me.compileExampleTimer.Tick, New System.EventHandler(AddressOf Me.leakSafeCompileEventRouter.OnCompileExampleTimerTick) 'OnCompileTimerTick
                Me.compileExampleTimer.Dispose()
                Me.compileExampleTimer = Nothing
            End If
        End Sub
    End Class

#Region "RichEditExampleEvaluatorByTimer"
    Public Class RichEditExampleEvaluatorByTimer
        Inherits RichEditDocumentServerAPIExample.CodeUtils.ExampleEvaluatorByTimer

        Public Sub New()
            MyBase.New()
        End Sub

        Protected Overrides Function GetExampleCodeEvaluator(ByVal language As RichEditDocumentServerAPIExample.CodeUtils.ExampleLanguage) As ExampleCodeEvaluator
            If language = RichEditDocumentServerAPIExample.CodeUtils.ExampleLanguage.VB Then Return New RichEditDocumentServerAPIExample.CodeUtils.RichEditVbExampleCodeEvaluator()
            Return New RichEditDocumentServerAPIExample.CodeUtils.RichEditCSExampleCodeEvaluator()
        End Function
    End Class

#End Region
#Region "LeakSafeCompileEventRouter"
    Public Class LeakSafeCompileEventRouter

        Private ReadOnly weakControlRef As System.WeakReference

        Public Sub New(ByVal [module] As RichEditDocumentServerAPIExample.CodeUtils.ExampleEvaluatorByTimer)
            'Guard.ArgumentNotNull(module, "module");
            Me.weakControlRef = New System.WeakReference([module])
        End Sub

        Public Sub OnCompileExampleTimerTick(ByVal sender As Object, ByVal e As System.EventArgs)
            Dim [module] As RichEditDocumentServerAPIExample.CodeUtils.ExampleEvaluatorByTimer = CType(Me.weakControlRef.Target, RichEditDocumentServerAPIExample.CodeUtils.ExampleEvaluatorByTimer)
            If [module] IsNot Nothing Then [module].CompileExample(sender, e)
        End Sub
    End Class

    Public Class CodeEvaluationEventArgs
        Inherits System.EventArgs

        Public Property Result As Boolean

        Public Property Code As String

        Public Property CodeClasses As String

        Public Property Language As ExampleLanguage

        Public Property EvaluationParameter As Object
    End Class

    Public Delegate Sub CodeEvaluationEventHandler(ByVal sender As Object, ByVal e As RichEditDocumentServerAPIExample.CodeUtils.CodeEvaluationEventArgs)

    Public Class OnAfterCompileEventArgs
        Inherits System.EventArgs

        Public Property Result As Boolean
    End Class

    Public Delegate Sub OnAfterCompileEventHandler(ByVal sender As Object, ByVal e As RichEditDocumentServerAPIExample.CodeUtils.OnAfterCompileEventArgs)
#End Region
End Namespace

File: CS/CodeUtils/Evaluator.cs
using System;
using System.CodeDom.Compiler;
using System.Reflection;

namespace RichEditDocumentServerAPIExample.CodeUtils
{
    public abstract class ExampleCodeEvaluator
    {
        protected abstract string CodeStart { get; }
        protected abstract string CodeBeforeClasses { get; }
        protected abstract string CodeEnd { get; }
        protected abstract CodeDomProvider GetCodeDomProvider();
        protected abstract string GetModuleAssembly();
        protected abstract string GetExampleClassName();

        public bool ExecuteCodeAndGenerateDocument(CodeEvaluationEventArgs args)
        {
            string theCode = String.Concat(CodeStart, args.Code, CodeBeforeClasses, args.CodeClasses, CodeEnd);
            string[] linesOfCode = new string[] { theCode };
            return CompileAndRun(linesOfCode, args.EvaluationParameter);
        }

        protected internal bool CompileAndRun(string[] linesOfCode, object evaluationParameter)
        {
            CompilerParameters CompilerParams = new CompilerParameters();

            CompilerParams.GenerateInMemory = true;
            CompilerParams.TreatWarningsAsErrors = false;
            CompilerParams.GenerateExecutable = false;

            string[] referencesSystem = new string[] { "System.dll",
                                                      "System.Windows.Forms.dll",
                                                      "System.Data.dll",
                                                      "System.Xml.dll",
                                                      "System.Core.dll",
                                                      "System.Drawing.dll" };

            string[] referencesDX = new string[] {
                AssemblyInfo.SRAssemblyData,
                GetModuleAssembly(),
                AssemblyInfo.SRAssemblyOfficeCore,
                AssemblyInfo.SRAssemblyPrintingCore,
                AssemblyInfo.SRAssemblyPrinting,
                AssemblyInfo.SRAssemblyDocs,
                AssemblyInfo.SRAssemblyUtils,
                AssemblyInfo.SRAssemblyRichEdit
            };
            string[] references = new string[referencesSystem.Length + referencesDX.Length];

            for (int referenceIndex = 0; referenceIndex < referencesSystem.Length; referenceIndex++)
            {
                references[referenceIndex] = referencesSystem[referenceIndex];
            }

            for (int i = 0, initial = referencesSystem.Length; i < referencesDX.Length; i++)
            {
                Assembly assembly = Assembly.Load(referencesDX[i] + AssemblyInfo.FullAssemblyVersionExtension);
                if (assembly != null)
                    references[i + initial] = assembly.Location;
            }
            CompilerParams.ReferencedAssemblies.AddRange(references);


            CodeDomProvider provider = GetCodeDomProvider();
            CompilerResults compile = provider.CompileAssemblyFromSource(CompilerParams, linesOfCode);

            if (compile.Errors.HasErrors)
            {
                string text = "Compile error: ";
                foreach (CompilerError ce in compile.Errors)
                {
                    text += "rn" + ce.ToString();
                }
                System.Windows.Forms.MessageBox.Show(text);
                return false;
            }

            Module module = null;
            try
            {
                module = compile.CompiledAssembly.GetModules()[0];
            }
            catch
            {
            }
            Type moduleType = null;
            if (module == null)
            {
                return false;
            }
            moduleType = module.GetType(GetExampleClassName());

            MethodInfo methInfo = null;
            if (moduleType == null)
            {
                return false;
            }
            methInfo = moduleType.GetMethod("Process");

            if (methInfo != null)
            {
                try
                {
                    methInfo.Invoke(null, new object[] { evaluationParameter });
                }
                catch (Exception)
                {
                    return false;// an error
                }
                return true;
            }
            return false;
        }
    }

    public abstract class RichEditExampleCodeEvaluator : ExampleCodeEvaluator
    {


        protected override string GetModuleAssembly()
        {
            return AssemblyInfo.SRAssemblyRichEditCore;
        }
        protected override string GetExampleClassName()
        {
            return "RichEditCodeResultViewer.ExampleItem";
        }
    }
    #region RichEditCSExampleCodeEvaluator
    public class RichEditCSExampleCodeEvaluator : RichEditExampleCodeEvaluator
    {

        protected override CodeDomProvider GetCodeDomProvider()
        {
            return new Microsoft.CSharp.CSharpCodeProvider();
        }
        const string codeStart =
      "using System;\r\n" +
            "using System.Text;\r\n" +
      "using DevExpress.XtraPrinting;\r\n" +
      "using DevExpress.XtraRichEdit;\r\n" +
      "using DevExpress.XtraRichEdit.API.Native;\r\n" +
            "using DevExpress.XtraRichEdit.Export;\r\n" +
            "using DevExpress.XtraRichEdit.Import;\r\n" +
            "using DevExpress.Office.Utils;\r\n" +
      "using System.Drawing;\r\n" +
      "using System.Windows.Forms;\r\n" +
      "using DevExpress.Utils;\r\n" +
      "using System.IO;\r\n" +
      "using System.Diagnostics;\r\n" +
      "using System.Xml;\r\n" +
      "using System.Data;\r\n" +
      "using System.Collections.Generic;\r\n" +
      "using System.Linq;\r\n" +
      "using System.Globalization;\r\n" +
      "using Document = DevExpress.XtraRichEdit.API.Native.Document;\r\n" +
      "using TableRow = DevExpress.XtraRichEdit.API.Native.TableRow;\r\n" +
      "namespace RichEditCodeResultViewer { \r\n" +
      "public class ExampleItem { \r\n" +
      "        public static void Process(RichEditDocumentServer wordProcessor) { \r\n" +
      "\r\n";

        const string codeBeforeClasses =
            "       \r\n }\r\n" +
            "    }\r\n";

        const string codeEnd =
            "\r\n" +
            "    }\r\n";

        protected override string CodeStart { get { return codeStart; } }
        protected override string CodeBeforeClasses
        {
            get { return codeBeforeClasses; }
        }
        protected override string CodeEnd { get { return codeEnd; } }
    }
    #endregion
    #region RichEditVbExampleCodeEvaluator
    public class RichEditVbExampleCodeEvaluator : RichEditExampleCodeEvaluator
    {

        protected override CodeDomProvider GetCodeDomProvider()
        {
            return new Microsoft.VisualBasic.VBCodeProvider();
        }
        const string codeStart =
      "Imports Microsoft.VisualBasic\r\n" +
      "Imports System\r\n" +
      "Imports DevExpress.XtraRichEdit\r\n" +
      "Imports DevExpress.XtraRichEdit.API.Native\r\n" +
      "Imports System.Drawing\r\n" +
      "Imports System.Windows.Forms\r\n" +
      "Imports DevExpress.Utils\r\n" +
      "Imports System.IO\r\n" +
      "Imports System.Diagnostics\r\n" +
      "Imports System.Xml\r\n" +
      "Imports System.Data\r\n" +
      "Imports System.Collections.Generic\r\n" +
      "Imports System.Globalization\r\n" +
      "Imports Document = DevExpress.XtraRichEdit.API.Native.Document\r\n"+
      "Imports TableRow = DevExpress.XtraRichEdit.API.Native.TableRow\r\n"+
      "Namespace RichEditCodeResultViewer\r\n" +
      "	Public Class ExampleItem\r\n" +
      "		Public Shared Sub Process(ByVal wordProcessor As RichEditDocumentServer)\r\n" +
      "\r\n";

        const string codeBeforeClasses =
            "\r\n		End Sub\r\n" +
            "	End Class\r\n";

        const string codeEnd =
        "\r\nEnd Namespace\r\n";

        protected override string CodeStart { get { return codeStart; } }
        protected override string CodeBeforeClasses
        {
            get { return codeBeforeClasses; }
        }
        protected override string CodeEnd { get { return codeEnd; } }
    }
    #endregion

    public abstract class ExampleEvaluatorByTimer : IDisposable
    {
        LeakSafeCompileEventRouter leakSafeCompileEventRouter;
        System.Windows.Forms.Timer compileExampleTimer;
        bool compileComplete = true;
        const int CompileTimeIntervalInMilliseconds = 2000;

        public ExampleEvaluatorByTimer(bool enableTimer)
        {
            this.leakSafeCompileEventRouter = new LeakSafeCompileEventRouter(this);

            if (enableTimer)
            {
                this.compileExampleTimer = new System.Windows.Forms.Timer();
                this.compileExampleTimer.Interval = CompileTimeIntervalInMilliseconds;

                this.compileExampleTimer.Tick += new EventHandler(leakSafeCompileEventRouter.OnCompileExampleTimerTick); //OnCompileTimerTick
                this.compileExampleTimer.Enabled = true;
            }
        }
        public ExampleEvaluatorByTimer()
            : this(true)
        {
        }

        #region Events
        public event CodeEvaluationEventHandler QueryEvaluate;

        protected internal virtual CodeEvaluationEventArgs RaiseQueryEvaluate()
        {
            if (QueryEvaluate != null)
            {
                CodeEvaluationEventArgs args = new CodeEvaluationEventArgs();
                QueryEvaluate(this, args);
                return args;
            }
            return null;
        }
        public event EventHandler OnBeforeCompile;

        void RaiseOnBeforeCompile()
        {
            if (OnBeforeCompile != null)
                OnBeforeCompile(this, new EventArgs());
        }

        public event OnAfterCompileEventHandler OnAfterCompile;

        void RaiseOnAfterCompile(bool result)
        {
            if (OnAfterCompile != null)
                OnAfterCompile(this, new OnAfterCompileEventArgs() { Result = result });
        }
        #endregion

        public void CompileExample(object sender, EventArgs e)
        {
            if (!compileComplete)
                return;
            CodeEvaluationEventArgs args = RaiseQueryEvaluate();
            if (!args.Result)
                return;

            ForceCompile(args);
        }
        public void ForceCompile(CodeEvaluationEventArgs args)
        {
            compileComplete = false;
            if (!String.IsNullOrEmpty(args.Code))
                CompileExampleAndShowPrintPreview(args);

            compileComplete = true;
        }
        void CompileExampleAndShowPrintPreview(CodeEvaluationEventArgs args)
        {
            bool evaluationSucceed = false;
            try
            {
                RaiseOnBeforeCompile();

                evaluationSucceed = Evaluate(args);
            }
            finally
            {
                RaiseOnAfterCompile(evaluationSucceed);
            }
        }

        public bool Evaluate(CodeEvaluationEventArgs args)
        {
            ExampleCodeEvaluator richeditExampleCodeEvaluator = GetExampleCodeEvaluator(args.Language);
            return richeditExampleCodeEvaluator.ExecuteCodeAndGenerateDocument(args);
        }

        protected abstract ExampleCodeEvaluator GetExampleCodeEvaluator(ExampleLanguage language);

        public void Dispose()
        {
            if (compileExampleTimer != null)
            {
                compileExampleTimer.Enabled = false;
                if (leakSafeCompileEventRouter != null)
                    compileExampleTimer.Tick -= new EventHandler(leakSafeCompileEventRouter.OnCompileExampleTimerTick); //OnCompileTimerTick
                compileExampleTimer.Dispose();
                compileExampleTimer = null;
            }
        }
    }

    #region RichEditExampleEvaluatorByTimer
    public class RichEditExampleEvaluatorByTimer : ExampleEvaluatorByTimer
    {
        public RichEditExampleEvaluatorByTimer()
            : base()
        {
        }

        protected override ExampleCodeEvaluator GetExampleCodeEvaluator(ExampleLanguage language)
        {
            if (language == ExampleLanguage.VB)
                return new RichEditVbExampleCodeEvaluator();
            return new RichEditCSExampleCodeEvaluator();
        }
    }
    #endregion

    #region LeakSafeCompileEventRouter
    public class LeakSafeCompileEventRouter
    {
        readonly WeakReference weakControlRef;

        public LeakSafeCompileEventRouter(ExampleEvaluatorByTimer module)
        {
            //Guard.ArgumentNotNull(module, "module");
            this.weakControlRef = new WeakReference(module);
        }
        public void OnCompileExampleTimerTick(object sender, EventArgs e)
        {
            ExampleEvaluatorByTimer module = (ExampleEvaluatorByTimer)weakControlRef.Target;
            if (module != null)
                module.CompileExample(sender, e);
        }
    }
    public class CodeEvaluationEventArgs : EventArgs
    {
        public bool Result { get; set; }
        public string Code { get; set; }
        public string CodeClasses { get; set; }
        public ExampleLanguage Language { get; set; }
        public object EvaluationParameter { get; set; }
    }
    public delegate void CodeEvaluationEventHandler(object sender, CodeEvaluationEventArgs e);

    public class OnAfterCompileEventArgs : EventArgs
    {
        public bool Result { get; set; }
    }
    public delegate void OnAfterCompileEventHandler(object sender, OnAfterCompileEventArgs e);
    #endregion

}
